### PR TITLE
Add messsage state to SB received message

### PR DESCRIPTION
**Pull request by rakshith91**
Fixes https://github.com/Azure/azure-sdk-for-python/issues/21634

### CODE REVIEW

File: sdk/servicebus/azure-servicebus/azure/servicebus/_common/constants.py [hunk 1]
```diff
@@ -165,12 +165,16 @@
...
+MESSAGE_STATE_NAME = b"x-opt-message-state"
 
 class ServiceBusReceiveMode(str, Enum):
     PEEK_LOCK = "peeklock"
     RECEIVE_AND_DELETE = "receiveanddelete"
 
+class ServiceBusMessageState(str, Enum):
+    ACTIVE = "Active"
+    DEFERRED = "Deferred"
+    SCHEDULED = "Scheduled"
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_common/constants.py:177]: I remember enum values should be of lower case?

```suggestion
    ACTIVE = "active"
    DEFERRED = "deferred"
    SCHEDULED = "scheduled"
```

File: sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py [hunk 2]
```diff
@@ -967,6 +969,27 @@ def dead_letter_source(self):
                 pass
         return None
 
+    @property
+    def message_state(self):
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py:973]: I was wondering should `ServiceBusMessage` also support this property with value always being `ACTIVE`?

could you check whether .NET just has the property for received message?

File: sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py [hunk 3]
```diff
@@ -967,6 +969,27 @@ def dead_letter_source(self):
...
+        # type: () -> Optional[str]
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py:974]: should it not be optional? as it always returns a status (defaults to `ACTIVE`)

or .NET returns None if it's undefined?

```suggestion
        # type: () -> ServiceBusMessageState
```

File: sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py [hunk 4]
```diff
@@ -967,6 +969,27 @@ def dead_letter_source(self):
...
+            try:
+                message_state = self._raw_amqp_message.annotations.get(MESSAGE_STATE_NAME)
+                return (
+                    ServiceBusMessageState.DEFERRED if message_state == 1
+                    else ServiceBusMessageState.SCHEDULED if message_state == 2
+                    else ServiceBusMessageState.ACTIVE
+                )
+            except AttributeError:
+                pass
+        return None
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py:991]: we check whether the annotations are available and if not, we return `None` which conflicts "Defaults to Active.".

Is this how .NET implement it as well?

File: sdk/servicebus/azure-servicebus/tests/async_tests/test_queues_async.py [hunk 5]
```diff
@@ -2123,3 +2124,53 @@ async def test_queue_async_send_amqp_annotated_message(self, servicebus_namespac
...
+            async with sender:
+                for i in range(10):
+                    message = ServiceBusMessage("message no. {}".format(i))
+                    await sender.send_messages(message)
+
+            receiver = sb_client.get_queue_receiver(servicebus_queue.name)
+            deferred_messages = []
+            async with receiver:
+                for message in receiver:
+                    deferred_messages.append(message.sequence_number)
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/tests/async_tests/test_queues_async.py:2171]: I think you could check whether message_state is ACTIVE before deferring, such that all enum states are covered

File: sdk/servicebus/azure-servicebus/tests/test_queues.py [hunk 6]
```diff
@@ -2581,3 +2583,49 @@ def test_queue_send_amqp_annotated_message(self, servicebus_namespace_connection
...
+            sender = sb_client.get_queue_sender(servicebus_queue.name)
+            for i in range(10):
+                message = ServiceBusMessage("message no. {}".format(i))
+                sender.send_messages(message)
+
+            receiver = sb_client.get_queue_receiver(servicebus_queue.name)
+            deferred_messages = []
+            with receiver:
+                for message in receiver:
+                    deferred_messages.append(message.sequence_number)
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/tests/test_queues.py:2627]: I think you could check whether message_state is ACTIVE before deferring, such that all enum states are covered

File: sdk/servicebus/azure-servicebus/CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,8 @@
 
 ### Features Added
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/CHANGELOG.md:5]: I remember this is a beta preview (correct me if I'm wrong), so remember to update the version number in change log/version.py, classifier in setup.py

File: sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py (hunk 2 above)
Comment (rakshith91) [reply sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py:973]: .net has it only in the received message  - also , we can add it later if necessary (but not remove it)

File: sdk/servicebus/azure-servicebus/tests/async_tests/test_queues_async.py [hunk 8]
```diff
@@ -2123,3 +2124,53 @@ async def test_queue_async_send_amqp_annotated_message(self, servicebus_namespac
...
                     assert recv_value_msg == 3
                     assert normal_msg == 4
+
+
+    @pytest.mark.liveTest
+    @pytest.mark.live_test_only
+    @CachedResourceGroupPreparer(name_prefix='servicebustest')
+    @CachedServiceBusNamespacePreparer(name_prefix='servicebustest')
+    @ServiceBusQueuePreparer(name_prefix='servicebustest', dead_lettering_on_message_expiration=True)
+    async def test_message_state_scheduled_async(self, servicebus_namespace_connection_string, servicebus_queue, **kwargs):
```
Comment (swathipil) [inline sdk/servicebus/azure-servicebus/tests/async_tests/test_queues_async.py:2134]: biggest baby nit:
I know the SB tests take a long time to run, and I'm not sure whether separating the message state tests into 2 makes them run longer?
If so, would it be possible to combine these two tests into one?
If not, ignore :)

Comment (rakshith91) [reply sdk/servicebus/azure-servicebus/tests/async_tests/test_queues_async.py:2134]: combined testing active and deferred - would like to separate scheduled and deferred for clarity purposes

File: sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py (hunk 2 above)
Comment (yunhaoling) [reply sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py:973]: sounds good!

File: sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py [hunk 9]
```diff
@@ -967,6 +969,27 @@ def dead_letter_source(self):
...
+            try:
+                message_state = self._raw_amqp_message.annotations.get(MESSAGE_STATE_NAME)
+                return (
+                    ServiceBusMessageState.DEFERRED if message_state == 1
+                    else ServiceBusMessageState.SCHEDULED if message_state == 2
+                    else ServiceBusMessageState.ACTIVE
+                )
+            except AttributeError:
+                pass
+        return ServiceBusMessageState.ACTIVE
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py:991]: I don't think we need the `if` check at the beginning? if self._raw_amqp_message.annotations is None, then `self._raw_amqp_message.annotations.get` triggers `AttributeError`, and we just return `ServiceBusMessageState.ACTIVE` in the except case

```suggestion
        try:
            message_state = self._raw_amqp_message.annotations.get(MESSAGE_STATE_NAME)
            return (
                ServiceBusMessageState.DEFERRED if message_state == 1
                else ServiceBusMessageState.SCHEDULED if message_state == 2
                else ServiceBusMessageState.ACTIVE
            )
        except AttributeError:
            return ServiceBusMessageState.ACTIVE
```

File: sdk/servicebus/azure-servicebus/CHANGELOG.md (hunk 7 above)
Comment (yunhaoling) [reply sdk/servicebus/azure-servicebus/CHANGELOG.md:5]: @rakshith91 , could you update the version number to be 7.6.0?

**Review by yunhaoling — APPROVED** [review]
LGTM!

**Pull request merged**